### ISSUE TITLE
Feature Request - Thermal Dynamics - Add a switch for fluxducts (and maybe liquiduct and itemduct)

### ISSUE THREAD

**Issue by Bjonnfesk** [CoFH/Feedback#940]
I wish to be able to turn off the power supply to certain machines on demand. I am aware that it is possible to do this with Energy Cells, but that is rather cumbersome and also sometimes requires you to use high-level Energy Cells, which is quite expensive. Therefore, I would much appreciate it if you could add a "Switch Fluxduct", or some other better name that I am sure you can come up with, to Thermal Dynamics. Perhaps make one for each tier of Fluxduct, and make switches for Itemducts and Liquiducts as well, because there have been situations where having that would have been useful for me. The switches should be turned on and off by Redstone signals, much like the EU-splitter cable from IndustrialCraft.

**Comment by Bjonnfesk** [comment]
A practical application where this would be extremely useful: I tend to use the Control Panel from Project Blue (unfortunately rare in modpacks, but I just install it as an extra mod) in conjunction with Wireless Redstone and Bundled Cable to control my entire base from a central hub, where I can turn on and off certain parts, such as the refinery and the quarries. Currently I am doing this by having an energy cell per machine/machine farm/factory/farm/whatever I wish to able to turn on and off, but like I said this is extremely cumbersome and requires me to use multiple Redstone Energy Cells which because a lot of my factories need that 8000RF/t output (which results in a lot of superfluous storage; in effect this becomes a sort of "inate capacitance" of the transmission lines). I then set the Redstone mode on the Energy Cells to be able to control them from the control panel. For things that are far away, like the quarries, I usually use Tesseracts which I turn on and off, but this is _even more_ cumbersome and expensive.

**Comment by KingLemming** [comment]
I understand the application, but we have to consider computational impacts as well. Splitting/recombining duct networks is fairly expensive, and simply creating a new block which is just a cell without the storage feels off.

**Comment by Bjonnfesk** [comment]
Indeed, I can imagine that would require some heavy recomputation. Possibly it could be added as a config setting that is default off, so that people who have powerful computers and/or can cope with any lag that may arise can turn it on if they wish.

However, I am not sure you understood me correctly (or possibly I am not understanding you correctly), since you use the words block and cell. I was trying to describe a sort of **duct** that would do this, not a block like a Redstone Energy Switch Cell that looks like a Redstone Energy Cell but has no storage. 

Or are you saying that having it as a **block** would be less computationally expensive for some reason that is too arcane to understand for people not well versed in the COFH code? I am a Java programmer myself, but I have not actually delved into the arts of Minecraft modding, so bear with me if you please.

**Comment by Bjonnfesk** [comment]
In fact, nevermind that, I understand what you mean now. I really cannot see how I failed to comprehend before, that was an utter derp on my part. New plan: would you be opposed to me having a go at making this myself? Not necessarily as a public mod, if you don't like the idea of that, but just for myself and a few friends that I play with.

**Comment by KingLemming** [comment]
I don't have any objection at all, but what you will basically end up creating is an Energy Cell without storage.

I can look into this, but just computationally it seems unwise to try it within the duct network itself.

**Comment by isochronous** [comment]
I am a software developer, but I have no personal experience with Minecraft mod development, so this may be a naive question, and if so I apologize... but would it really be necessary to recalculate the whole duct network every time a "switch" (or whatever) gets toggled? It seems like you should basically be able to calculate all of the computationally heavy stuff when the ducts are placed and just cache the results. If a network includes a switch then you do three calculations - one for both segments together, and one for each of the other two networks. Things would get more difficult when you have more than one switch, but it should still be doable. I'm kinda thinking that every switch would cache the necessary data for both/all of its sides... that is to say, once routing calculations are performed for one side of a switch, the results are cached somewhere that the switch can access them. When a switch is opened (stuff flowing through), the cached results are accessed and the "new" part of the network is treated as just one large component that you already have the calculations for. When it's closed, it's like removing that same component.

I don't think I'm doing a good job of explaining what's in my head, but if there's any validity to the idea I'm sure you can pick it up and run with it.

Alternatively, why not create an energy cell without storage? I mean, if you called it that, then yeah, it would seem dumb, but the way OP is describing his setup, it sounds like what he wants is basically the equivalent of a switching diode. You could create a "flux duct switch" for each tier that's literally an energy cell with no storage code-wise (maybe even same model but different texture), with no configurable input/output rate, as the transfer rate would be exactly the same as the flux duct of the same tier. It wouldn't be any more computationally heavy than using an energy cell, could be significantly cheaper, and could even get unique upgrades like being able to switch between multiple input/output configurations based on the strength of redstone signal applied. I know I've had more than one use case where something like that could have come in very handy.

**Comment by skyboy** [comment]
@isochronous Adding it as part of the network logic itself would require splitting and rejoining networks each time the device is toggled, while joining two networks is fast enough, splitting them requires traversing the in-world blocks to re-construct the network. There is some possibility of caching the locations of all the constituent blocks in a separate data structure (hextree?) for faster re-traversal, but that runs into problems with frame mods and other such crap that nothing is really designed to handle.

**Comment by isochronous** [comment]
Oh yeah, I completely forgot about frames... that would present a problem.

**Comment by Phebus666** [comment]
A compromise could be to have the option to limit the capacity of energy cells, same way we can limit throughput. That would at least help with the "innate capacity" issue.

**Comment by bookerthegeek** [comment]
Even better, why not add a servo that acts to the network like a zero capacity cell?

The duct network would see that final pipe now as an energy cell with zero storage, even though it is in all appearance a duct with a servo on it.

This would allow redstone control without the entire network needing to be rebuilt when switched, and I would think that it would be no more computational expensive then having an energy cell after the duct.

**Comment by gvlfm78** [comment]
:+1: for this idea
